py/src/braintrust/logger.py

```diff
@@ -1121,6 +1147,65 @@ def __exit__(self, type, value, callback):
         self.end()
 
 
+class _ObjectLoader:
+    def __init__(self, object_type: str, object_id: id, version: "str | int" = None):
+        self.finished = False
+        self.object_type = object_type
+        self.object_id = object_id
+
+        self._fetched_data = None
+
+        self._pinned_version = None
+        if version is not None:
+            try:
+                self._pinned_version = int(version)
+                assert self._pinned_version >= 0
+            except (ValueError, AssertionError):
+                raise ValueError(f"version ({version}) must be a positive integer")
+
+    def fetch(self):
+        self._check_not_finished()
+
+        for record in self.fetched_data:
+            yield {
+                "id": record.get("id"),
+                "input": json.loads(record.get("input") or "null"),
+                "output": json.loads(record.get("output") or "null"),
+                "metadata": json.loads(record.get("metadata") or "null"),
+            }
+
+        self._clear_cache()
```

Comment:
I understand this is just moving code around, but I find the two methods `fetch` and `fetched_data` a bit confusing. They both yield similar outputs (not exactly the same), and one clears the cache while the other doesn't. What's the rationale?